depinject/provider_desc.go

```diff
@@ -25,85 +28,94 @@ type ProviderDescriptor struct {
 	Location Location
 }
 
-type ProviderInput struct {
+type providerInput struct {
 	Type     reflect.Type
 	Optional bool
 }
 
-type ProviderOutput struct {
+type providerOutput struct {
 	Type reflect.Type
 }
 
-func ExtractProviderDescriptor(provider interface{}) (ProviderDescriptor, error) {
-	rctr, ok := provider.(ProviderDescriptor)
-	if !ok {
-		var err error
-		rctr, err = doExtractProviderDescriptor(provider)
-		if err != nil {
-			return ProviderDescriptor{}, err
-		}
+func extractProviderDescriptor(provider interface{}) (providerDescriptor, error) {
+	rctr, err := doExtractProviderDescriptor(provider)
+	if err != nil {
+		return providerDescriptor{}, err
 	}
-
-	return expandStructArgsProvider(rctr)
+	return postProcessProvider(rctr)
 }
 
-func ExtractInvokerDescriptor(provider interface{}) (ProviderDescriptor, error) {
-	rctr, ok := provider.(ProviderDescriptor)
-	if !ok {
-		var err error
-		rctr, err = doExtractProviderDescriptor(provider)
-
-		// mark all inputs as optional
-		for i, input := range rctr.Inputs {
-			input.Optional = true
-			rctr.Inputs[i] = input
-		}
+func extractInvokerDescriptor(provider interface{}) (providerDescriptor, error) {
+	rctr, err := doExtractProviderDescriptor(provider)
+	if err != nil {
+		return providerDescriptor{}, err
+	}
 
-		if err != nil {
-			return ProviderDescriptor{}, err
-		}
+	// mark all inputs as optional
+	for i, input := range rctr.Inputs {
+		input.Optional = true
+		rctr.Inputs[i] = input
 	}
 
-	return expandStructArgsProvider(rctr)
+	return postProcessProvider(rctr)
 }
 
-func doExtractProviderDescriptor(ctr interface{}) (ProviderDescriptor, error) {
+func doExtractProviderDescriptor(ctr interface{}) (providerDescriptor, error) {
 	val := reflect.ValueOf(ctr)
 	typ := val.Type()
 	if typ.Kind() != reflect.Func {
-		return ProviderDescriptor{}, errors.Errorf("expected a Func type, got %v", typ)
+		return providerDescriptor{}, errors.Errorf("expected a Func type, got %v", typ)
 	}
 
-	loc := LocationFromPC(val.Pointer())
+	loc := LocationFromPC(val.Pointer()).(*location)
```

Comment:
Exported Providers and Invokers are a requirement for codegen, did you consider allowing them for them for runtime DI? Maybe this would be too confusing for users of this lib though.

Reply:
I think you mean allowing unexported providers at runtime. I think it would be too confusing to have different behavior for runtime and compile time. Also, we want to ensure that all code can be executed at compile time if desired so doing this check at runtime ensures that.